feat: Add toHaveValue matcher #82

lukaszfiszer opened the pull request:
**What**:
Add `toHaveValue` matcher

**Why**:
As discussed in #82 

**How**:
Reused a lot of existing logic from `toHaveFormValue` and moved some internal functions  to `utils.js` so that they can be shared between both modules.

I had a lot of thoughts on how to implement this matcher for `<input type="checkbox" />` and `<input type="radio" />`, because those elements value can be different depending on other elements from the same group (sharing the same `name`). Finally I've decided to throw an error when a checkbox or radio input is passed, suggesting users to use `toHaveFormValue` instead. In the future, I think I would most make sense to implement `toBeChecked` matcher - a perfect pair with to `toHaveValue` for single form elements. 

**Checklist**:

* [x] Documentation
* [x] Tests
* [x] Updated Type Definitions
* [x] Ready to be merged <!-- In your opinion, is this ready to be merged as soon as it's reviewed? -->
* [x] Added myself to contributors table <!-- this is optional, see the contributing guidelines for instructions -->

gnapse commented on README.md:
```diff
@@ -687,6 +688,63 @@ expect(element).not.toHaveTextContent('content')
+This allows you to check whether the given form element has the specified value.
+It accepts `<input>`, `<select>` and `<textarea>` elements with the exception of
+of `<input type="checkbox">` and `<input type="radio">`, which can be meaningfully
+matched only using [`toHaveFromValue`](#tohaveformvalues).
```
Typo `toHaveFromValue` should be `toHaveFormValue`

gnapse commented on README.md:
```diff
@@ -687,6 +688,63 @@ expect(element).not.toHaveTextContent('content')
+matched only using [`toHaveFromValue`](#tohaveformvalues).
+
+For all other form elements, the value is matched using the same algorithm
+as in [`toHaveFromValue`](#tohaveformvalues) does.
```
Same typo.

lukaszfiszer commented:
@gnapse I have fix the typos.

Another idea that I had for checkbox and radio inputs was to allow the user to pass a parent element of those elements:

```html
<div data-testid="contact">
   <label><input type="radio" name="contact" value="email">Email</label>
   <label><input type="radio" name="contact" value="phone">Phone</label>
   <label><input type="radio" name="contact" value="post">Post</label>
</div>
```

```js
expect(getByTestId('contact')).toHaveValue('phone')
```

But then probably you could  just use `toHaveFormValues` for that.

Let me know what you think

kentcdodds commented:
What do we need to do to help get this across the finish line?

gnapse commented:
I probably forgot to keep track of this and approve or request changes at the time. But now it seems there are conflicts, and also the build is failing for some reason (could it have something to do with the change of url of the github repo when moved to an org @kentcdodds?  When I visit the travis details link above I get "We couldn't find the repository gnapse/jest-dom").

@lukaszfiszer sorry about not being on top of this. Can you please fix the conflicts? We'll figure out the travis issue in the mean time.

lukaszfiszer commented:
@gnapse No worries. I will fix the conflicts over the weekend.

lukaszfiszer commented:
@gnapse rebased against master and fixed the broken test. ready to merge

jgoz commented on extend-expect.d.ts:
```diff
@@ -20,5 +20,6 @@ declare namespace jest {
       text: string | RegExp,
       options?: {normalizeWhitespace: boolean},
     ): R
+    toHaveValue(value?: string | string[]): R
```
Should this be `value?: string | string[] | number`?

jgoz commented on README.md:
````diff
@@ -690,6 +691,63 @@ expect(element).not.toHaveTextContent('content')
+### `toHaveValue`
+
+```typescript
+toHaveValue(value: string | string[])
````
And should this one be `value?: string | string[] | number`?

jgoz commented on src/__tests__/to-have-value.js:
```diff
@@ -0,0 +1,88 @@
+    `)
+
+    expect(queryByTestId('with')).toHaveValue('foo')
+    expect(queryByTestId('with')).not.toHaveValue('bar')
```
Would it make sense to add a test case like this?

```js
expect(queryByTestId('with')).toHaveValue();
```

The corresponding `.not.toHaveValue()` case is already there for `empty`.

lukaszfiszer commented:
@jgoz Thanks for review. I've fixed the typings, added additional test cases and fixed the behaviour of `.toHaveValue()` (without argument) - it asserts if input/select has *any* value.

jgoz commented on src/index.js:
```diff
@@ -29,7 +30,9 @@ export {
   toBeRequired,
   toBeInvalid,
   toBeValid,
+  toHaveValue
```
This is being exported twice

jgoz commented on src/to-have-value.js:
```diff
@@ -0,0 +1,44 @@
+    ['checkbox', 'radio'].includes(htmlElement.type)
+  ) {
+    throw new Error(
+      'input with type=checkbox or type=radio cannot be used with .toHaveValue(). Use .toHaveFormValue() instead',
```
Minor: `.toHaveFormValue()` should be `.toHaveFormValues()`

jgoz commented:
@lukaszfiszer sorry for the delayed response, I missed the notification on this. Just two very minor issues that I can see, one of which is causing the CI failure.

lukaszfiszer commented:
@jgoz fixed issues and squash-rebased against master. Hope to get it merged!

kentcdodds commented on README.md:
````diff
@@ -839,6 +840,63 @@ expect(element).not.toHaveTextContent('content')
+```javascript
+const {getByTestId} = render(/* Rendered HTML */)
+
+const textInput = getByTestId('[data-testid="input-text"]')
````
I think these should be: `getByTestId('input-text')` and so-forth.

lukaszfiszer replied:
Good catch! Fixed.

gnapse commented:
:tada: This PR is included in version 3.5.0 :tada:

The release is available on:
- `npm package (@latest dist-tag)`
- [GitHub release](https://github.com/testing-library/jest-dom/releases/tag/v3.5.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
